### ISSUE TITLE
C preprocessor code-generator

### ISSUE THREAD

**Issue by xor-gate** [ludocode/mpack#9]
Hi Nicholas,

I wanted to let you know how to abuse the C preprocessor to generate C-code to serialize C-structs into MessagePack. This uses the "X-Macros" concept using the CPP from https://en.wikibooks.org/wiki/C_Programming/Preprocessor#X-Macros. 

output:
```
name [string_t]: (null)
x [uint8_t]: 0
y [uint8_t]: 0
z [uint8_t]: 0
name [string_t]: Sirius
x [uint8_t]: 255
y [uint8_t]: 127
z [uint8_t]: 63
MessagePack: [84 a4 6e 61 6d 65 a6 53 69 72 69 75 73 a1 78 cc ff a1 79 7f a1 7a 3f ]
```

```c
/*
http://msgpack-json-editor.com/
https://en.wikibooks.org/wiki/C_Programming/Preprocessor#X-Macros
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <inttypes.h>

#include "mpack-config.h"
#include <mpack.h>

/*
 Generic
 */
#define STRUCT_MEMBER(member, type, dummy) type member;

#define SERIALIZE_KEY(writer, key) mpack_write_cstr(writer, #key);
#define SERIALIZE_VALUE(writer, value, type) SERIALIZE_##type(writer, value)
#define SERIALIZE_uint8_t(writer, value)  mpack_write_u8(writer, value)
#define SERIALIZE_string_t(writer, value) mpack_write_cstr(writer, value);

#define SERIALIZE_MEMBER(member, type, obj, writer) \
	SERIALIZE_KEY(writer, member); \
	SERIALIZE_VALUE(writer, obj->member, type);

#define DESERIALIZE_MEMBER(member, type, obj, buffer) \
  memcpy(&(obj->member), buffer, sizeof(obj->member)); \
  buffer += sizeof(obj->member);

#define FORMAT_(type) FORMAT_##type
#define FORMAT_int    "%d"
#define FORMAT_double "%g"
#define FORMAT_uint8_t "%" PRIu8
#define FORMAT_string_t "%s"

/* FORMAT_(type) will be replaced with FORMAT_int or FORMAT_double */
#define PRINT_MEMBER(member, type, obj) \
  printf("%s [%s]: " FORMAT_(type) "\n", #member, #type, obj->member);

typedef const char * string_t;

/*
 starStruct
 */

#define EXPAND_STAR(_, ...) \
  _(name, string_t, __VA_ARGS__) \
  _(x, uint8_t, __VA_ARGS__) \
  _(y, uint8_t, __VA_ARGS__) \
  _(z, uint8_t, __VA_ARGS__)

typedef struct {
  EXPAND_STAR(STRUCT_MEMBER, )
  } starStruct;

void serialize_star(const starStruct *const star, mpack_writer_t *writer) {
  mpack_start_map(writer, 4);
  EXPAND_STAR(SERIALIZE_MEMBER, star, writer)
  mpack_finish_map(writer);
  }

void deserialize_star(starStruct *const star, const unsigned char *buffer) {
  EXPAND_STAR(DESERIALIZE_MEMBER, star, buffer)
  }

void print_star(const starStruct *const star) {
  EXPAND_STAR(PRINT_MEMBER, star)
  }

int
main(void)
{
	char* data;
	size_t size;
	static mpack_writer_t writer;

	starStruct s;
	memset(&s, 0, sizeof(s));
	print_star(&s);

	s.name = "Sirius";
	s.x = 255;
	s.y = 127;
	s.z = 63;

	print_star(&s);

	// encode to memory buffer
	mpack_writer_init_growable(&writer, &data, &size);

	// write start to msgpack
	serialize_star(&s, &writer);

	// finish writing
	if (mpack_writer_destroy(&writer) != mpack_ok) {
		fprintf(stderr, "An error occurred encoding the data!\n");
		return -1;
	}

	printf("MessagePack: [");
	for(unsigned int j = 0; j < mpack_writer_buffer_used(&writer); j++)
		printf("%02x ", ((uint8_t *)data)[j]);
	printf("]\n");

	// use the data
	free(data);

	return 0;
}
```

This is just a example how to use the preprocessor to "map" single-depth C-structs onto MessagePack without writing any script at all. It is still a very crude way but for Embedded Firmware purposes it SHOULD work on every compiler. And deserialize is not yet covered using mpack.

Kind regards,
Jerry

**Comment by ludocode** [comment]
Sorry, I did mean to comment on this at some point. This is a neat way to generate helpers for the expect API to map MessagePack data to a struct. I had thought for a while if there was a good way with macros, and this looks somewhat workable.

I was thinking a good direction to go with this would be to write a simple script that parses some sort of schema file and generates the MPack Expect and Write code, just like Protocol Buffers, FlatBuffers, Apache Thrift, etc. Of course if you're willing to do that, you might as well just use FlatBuffers. I believe there are huge advantantages to having a typed schemaless underlying format though, so if something like this existed I would likely still want to use it rather than protobuf/flatbuffers even if it's a bit less efficient. I believe MessagePack is incredibly efficient despite the type information, so I'm tempted to write something like this if only to try to beat Protocol Buffers in performance :) Anyway I've got way too many other things on my todo list first, so this would be a long way down the road.
